### PR TITLE
closed #450 and #197: handling big integers and better preprocessing

### LINKED ISSUE #450: [FEATURE] Make sure that apalache works with math over big integers
SMT does not restrict the size of the integers. Neither should do apalache. We have to make sure that the preprocessing does not accidentally truncate big integers.

### PR DESCRIPTION
**Pull request by konnov**
This PR reworks the constant preprocessor to:
  - fix operations on big integers: #450 
  - fix preprocessing of string comparisons: #197 

As a bonus, we can check Solidity's [SafeMath](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/math/SafeMath.sol), see [SafeMath.tla](https://github.com/informalsystems/apalache/blob/igor/bigint450/test/tla/SafeMath.tla). And the results are intriguing :-)

`ConstSimplifier` and `ConstSimplifierForSmt` need a massive set of unit tests. Postponed for #451 .

- [ ] ~Tests added for any new code~
- [ ] ~Documentation added for any new functionality~
- [X] Entry added to [UNRELEASED.md](./UNRELEASED.md) for any new functionality

### CODE REVIEW

File: test/tla/SafeMath.tla [hunk 1]
```diff
@@ -0,0 +1,118 @@
+---------------------------- MODULE SafeMath -----------------------------------
+(*
+ * Test that the operations over big integers go through.
+ * This specification is expected to deadlock, but not violating the invariant.
```
Comment (shonfeder) [inline test/tla/SafeMath.tla:4]: For mysterious reasons of "idiomatic" sound, I suggest

```suggestion
 * This specification is expected to deadlock, but not violate the invariant.
```

or

```suggestion
 * This specification is expected to deadlock, without violating the invariant.
```

File: test/tla/SafeMath.tla [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+          (*
+          \/ TestTry("MUL", TryMul)
+          \/ TestTry("DIV", TryDiv)
+          \/ TestTry("MOD", TryMod)
+            *)
+    \/ opcode /= "NOP" /\ UNCHANGED vars
+
+InvAdd ==
+    opcode = "ADD"
+        => is_error <=> (arg1 + arg2 > MAX_UNSIGNED /\ res /= arg1 + arg2)
```
Comment (shonfeder) [inline test/tla/SafeMath.tla:94]: In these invariants, for readability sake (and maybe for performance, depending on how the operations are handled), would it make sense to do the operation once via `LET` binding?

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rewriter/ConstSimplifierForSmt.scala [hunk 3]
```diff
@@ -5,14 +5,17 @@ import at.forsyte.apalache.tla.bmcmt.smt.SolverContext
...
 import at.forsyte.apalache.tla.lir.values.{TlaBool, TlaInt}
+import at.forsyte.apalache.tla.pp.ConstSimplifierBase
 
 /**
-  * A simplifier of constant TLA+ expressions, e.g., rewriting 1 + 2 to 3.
-  * This simplifier is using some knowledge about SMT.
+  * <p>A simplifier of constant TLA+ expressions, e.g., rewriting 1 + 2 to 3.
+  * This simplifier is using some knowledge about SMT.</p>
+  *
+  * <p>Bugfix #450: make sure that the integers are simplified with BigInt.</p>
```
Comment (shonfeder) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rewriter/ConstSimplifierForSmt.scala:14]: I'm not sure I see the value of putting bugfix history in the documentation. Is this common practice in Java/Scala land?

**Review by shonfeder — APPROVED** [review]

File: test/tla/SafeMath.tla (hunk 2 above)
Comment (konnov) [reply test/tla/SafeMath.tla:94]: Apalache will cache the expressions, but let's add the LET-definition for readability

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rewriter/ConstSimplifierForSmt.scala (hunk 3 above)
Comment (konnov) [reply tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rewriter/ConstSimplifierForSmt.scala:14]: Well, I don't put every bugfix in the code. Sometimes, I add them, not to repeat the same mistake twice :-)

**Pull request merged**